registry/cache.go

```diff
@@ -14,28 +14,32 @@ import (
 )
 
 type Cache struct {
-	backend
+	next   dockerRegistryInterface
 	creds  Credentials
 	expiry time.Duration
 	Client MemcacheClient
 	logger log.Logger
 }
 
-func NewCache(b backend, creds Credentials, cache MemcacheClient, expiry time.Duration, logger log.Logger) *Cache {
-	return &Cache{
-		backend: b,
-		creds:   creds,
-		expiry:  expiry,
-		Client:  cache,
-		logger:  logger,
+type CachedDockerRegistry func(dockerRegistryInterface) dockerRegistryInterface
+
+func NewCache(creds Credentials, cache MemcacheClient, expiry time.Duration, logger log.Logger) CachedDockerRegistry {
```

Comment:
All the metrics stuff uses it. And all the http stuff uses it. Google golang middleware and you can read for days. And GoKit uses it for nearly everything. From logging to rate limiting.